File: src/commands/file/encrypt.ts
import {DocumentAccessList, DocumentIDNameResponse} from "@ironcorelabs/ironnode";
import {Command, flags as flagtype} from "@oclif/command";
import * as fs from "fs";
import * as GroupMaps from "../../lib/GroupMaps";
import {ironnode} from "../../lib/SDK";
import {keyFile} from "../../lib/sharedFlags";
import * as Utils from "../../lib/Utils";
import chalk = require("chalk");

type EncryptResult = Utils.ErrorOr<DocumentIDNameResponse>;

const FILE_SIZE_ERROR_THRESHOLD = 1024 * 1024 * 1024 * 2; //2 GB

/**
 * File encrypt command. Takes a single file or multiple files and encrypts them to the caller and any optional users and
 * groups provided via flags. Multiple files and single files without the -o flag will write a file as a sibling of the plaintext
 * file with a `.iron` suffix. Providing the -o option allows users to write files to a specific file location or to use `-` to
 * write the encrypted bytes to stdout.
 */
export default class Encrypt extends Command {
    static description =
        "Encrypt a file or list of files to yourself and optionally to other users or groups. By default, the input file is unchanged and the output uses the same filename with a '.iron' extension added.";
    static strict = false; //Allow multiple files to be provided as arguments
    static args = [
        {
            name: "files",
            description: "Path of file or files to encrypt.",
        },
    ];
    static flags = {
        help: flagtype.help({char: "h"}),
        keyfile: keyFile(),
        users: flagtype.option({
            char: "u",
            description: "Encrypt the file(s) to a comma-separated list of user emails. Files are automatically encrypted to the logged-in user.",
            parse: (list) => list.split(","),
        }),
        groups: flagtype.option({
            char: "g",
            description: "Encrypt the file(s) to a comma-separated list of groups.",
            parse: (list) => list.split(","),
        }),
        out: flagtype.string({
            char: "o",
            description:
                "Filename where encrypted file will be written. Only allowed if a single file is being encrypted. Use '-o -' to write encrypted file content to stdout, but fair warning, the output is binary and not ASCII.",
        }),
        stdin: flagtype.boolean({
            char: "s",
            description: "Read data to encrypt from stdin. If used, no source files should be provided as arguments and you must use the '-o' flag.",
            dependsOn: ["out"],
        }),
        delete: flagtype.boolean({
            char: "d",
            description: "Delete the unencrypted source file(s) after successful encryption.",
            exclusive: ["stdin"],
        }),
    };
    static examples = [
        Utils.buildCommandSampleText("file:encrypt path/to/file", "Encrypt the provided file and write the results to 'path/to/file.iron'."),
        Utils.buildCommandSampleText(
            "file:encrypt -g myGroup -u john@example.com,mike@example.com path/to/file",
            "Encrypt the provided file and grant decryption access to two users and one group"
        ),
        Utils.buildCommandSampleText(
            "file:encrypt path/to/file -o other/path/to/encrypted.iron",
            "Encrypt the provided file and write the decrypted bytes to 'other/path/to/encrypted.iron'."
        ),
        Utils.buildCommandSampleText(
            "file:encrypt path/to/file.txt -d",
            "Encrypt the provided file and write the decrypted bytes to 'path/to/file.txt.iron' while also deleting the original unencrypted 'path/to/file.txt' file."
        ),
        Utils.buildCommandSampleText("file:encrypt *.json", "Encrypt all of the JSON files in the current directory and write them out to '.iron' files."),
        Utils.buildCommandSampleText("file:encrypt path/to/file -o -", "Encrypt the provided file and write the encrypted content to stdout."),
        Utils.buildCommandSampleText(
            `${chalk.green("echo")} "my secret" | ${chalk.green("ironhide")} file:encrypt -s -o - | ${chalk.green("base64")} -e`,
            "Encrypt stdin and write the encrypted results to stdout in base64.",
            true
        ),
    ];

    /**
     * Get the readable stream that the content to encrypt is coming from. Either uses the process.stdin stream or verifies that the
     * provided sourcePath has the proper permissions and creates a readable stream from it.
     */
    getSourceStream(encryptOp: Utils.ProcessSingleFileOp | Utils.ProcessStdinOp): Promise<NodeJS.ReadableStream> {
        if (Utils.isStdInFileOperation(encryptOp)) {
            return Promise.resolve(process.stdin);
        }
        const sourceFile = Utils.normalizePathToFile(encryptOp.file);
        return Utils.checkSourceFilePermissions(sourceFile).then(() => {
            const sourceFileStats = fs.lstatSync(sourceFile);
            if (sourceFileStats.size > FILE_SIZE_ERROR_THRESHOLD) {
                return Promise.reject(new Error(`${sourceFile} is over 2GB, which is not currently supported.`));
            }
            return Promise.resolve(fs.createReadStream(sourceFile));
        });
    }

    /**
     * Get the writable stream that the encrypted content will be piped out to. Either uses process.stdout stream or creates a writable
     * stream from either the outputFlag path or the sourcePath with a .iron extension.
     */
    getDestinationStream(encryptOp: Utils.ProcessSingleFileOp | Utils.ProcessStdinOp): Promise<NodeJS.WritableStream> {
        if (encryptOp.out === "-") {
            return Promise.resolve(process.stdout);
        }
        const destinationFile = Utils.normalizePathToFile(encryptOp.out || `${(encryptOp as Utils.ProcessSingleFileOp).file}.iron`);
        return Utils.checkDestinationFilePermissions(destinationFile).then(() => fs.createWriteStream(destinationFile));
    }

    /**
     * Run both operations to get the source and destinations that we'll be reading/writing from for this operation.
     */
    getSourceAndDestinationStreams(encryptOp: Utils.ProcessSingleFileOp | Utils.ProcessStdinOp) {
        return this.getSourceStream(encryptOp).then((readStream) => this.getDestinationStream(encryptOp).then((writeStream) => [readStream, writeStream]));
    }

    /**
     * Returns a promise for the actual encrypt operation. Determines the source and destination streams and returns a Promise which runs
     * the encrypt operation. NOTE: The returned Promise from this method CANNOT fail as we let the caller decide how to handle failure. This
     * lets us prevent a complete failure when encrypting multiple files when one of them fails.
     */
    getFileEncryptPromise(fileAccessList: DocumentAccessList, encryptOp: Utils.ProcessSingleFileOp | Utils.ProcessStdinOp): Promise<EncryptResult> {
        return this.getSourceAndDestinationStreams(encryptOp)
            .then(([sourceStream, destStream]) =>
                ironnode().document.encryptStream(sourceStream as NodeJS.ReadStream, destStream as NodeJS.WriteStream, {accessList: fileAccessList})
            )
            .then((decryptResult) => {
                if (!Utils.isStdInFileOperation(encryptOp) && encryptOp.deleteSource) {
                    try {
                        Utils.deleteFile(encryptOp.file);
                    } catch (_) {
                        this.warn(`Unable to delete source file '${encryptOp.file}' as it is not writable.`);
                    }
                }
                return decryptResult;
            })
            .catch((e) => e);
    }

    /**
     * Encrypt a list of files to the provided list of users and groups. Runs a series of Promises in parallel and logs the results of the successes
     * and failures.
     */
    encryptMultipleFiles(fileAccessList: DocumentAccessList, encryptOp: Utils.ProcessMultipleFileOp) {
        const encryptOps = encryptOp.file.map((file) =>
            this.getFileEncryptPromise(fileAccessList, {file, out: encryptOp.out, deleteSource: encryptOp.deleteSource})
        );

        return Promise.all(encryptOps).then((results) => {
            let successfulCount = 0;
            let failureCount = 0;
            results.forEach((encryptionResult) => {
                if (Utils.isError(encryptionResult)) {
                    this.log(chalk.red(encryptionResult.message));
                    failureCount++;
                } else {
                    successfulCount++;
                }
            });
            if (successfulCount > 0) {
                this.log(chalk.green(`\n${successfulCount} file(s) successfully encrypted.`));
            }
            if (failureCount > 0) {
                this.error(chalk.red(`\n${failureCount} file(s) failed to be encrypted.`));
            }
        });
    }

    /**
     * Encrypt a single file from it's path or from stdin and write the output to either
     * + A sibling of the source file with a .iron extension,
     * + The file path provided by the -o flag
     * + Stdout if the -o flag was set to '-'
     */
    encryptSingleFile(fileAccessList: DocumentAccessList, encryptOp: Utils.ProcessSingleFileOp | Utils.ProcessStdinOp) {
        return this.getFileEncryptPromise(fileAccessList, encryptOp).then((encryptionResult) => {
            if (Utils.isError(encryptionResult)) {
                //Throw this error so that it's caught in the try/catch in the run() function so we log this as an error
                throw encryptionResult;
            }
            if (encryptOp.out !== "-") {
                const writtenResult = Utils.isStdInFileOperation(encryptOp) ? encryptOp.out : encryptOp.out || `${encryptOp.file}.iron`;
                this.log(chalk.green(`Encrypted file successfully written to ${writtenResult}.`));
            }
        });
    }

    /**
     * Various argument/outflag/stdinflag checks to verify that we support the combination of things they're trying to do.
     */
    checkFlagsAndArguments(argv: string[], outFlag?: string, stdinFlag?: boolean) {
        if (argv.length > 1 && outFlag) {
            return this.error(chalk.red("The output flag (-o) can only be provided when encrypting a single file."));
        }
        if (argv.length === 0 && !stdinFlag) {
            return this.error(chalk.red("You must either provide a path to a file to encrypt or use the -s flag to read from stdin."));
        }
        if (argv.length > 0 && stdinFlag) {
            return this.error(chalk.red("You cannot provide both source files and the -s flag."));
        }
    }

    async run() {
        const {argv, args, flags} = this.parse(Encrypt);
        this.checkFlagsAndArguments(argv, flags.out, flags.stdin);
        let requestedGroups: string[] = [];
        //Only lookup the users groups for mapping if they're sharing with a group
        if (flags.groups && flags.groups.length) {
            const [groupsByName] = await GroupMaps.getGroupMaps();
            requestedGroups = await GroupMaps.convertGroupNamesToIDs(flags.groups, groupsByName);
        }

        const accessList = Utils.convertUserAndGroupToAccessList(flags.users, requestedGroups);

        let encryptOp: Utils.ProcessOp;
        if (flags.stdin) {
            encryptOp = {out: flags.out as string} as Utils.ProcessStdinOp;
        } else {
            encryptOp = {
                file: argv.length > 1 ? argv : (args.files as string),
                out: flags.out,
                deleteSource: flags.delete,
            } as Utils.ProcessFileOp;
        }

        try {
            Utils.isMultipleFileOperation(encryptOp)
                ? await this.encryptMultipleFiles(accessList, encryptOp)
                : await this.encryptSingleFile(accessList, encryptOp);
        } catch (e) {
            this.error(chalk.red(e.message));
        }
    }
}

File: src/file/encrypt.rs
use crate::{
    group_maps::{convert_group_names_to_ids, get_group_maps},
    util::{self, act_on_all_files},
};
use clap::Parser;
use ironoxide::prelude::*;
use itertools::Itertools;
use itertools::{Either, EitherOrBoth};
use prettytable::Row;
use std::{
    convert::TryFrom,
    fs::{self, OpenOptions},
    io::{self, Read, Write},
    path::PathBuf,
};
use yansi::Paint;

const EXAMPLES: &str = "EXAMPLES

    Encrypt the provided file and write the results to 'path/to/file.iron'.
        $ ironhide file encrypt path/to/file

    Encrypt the provided file and grant decryption access to two users and one group
        $ ironhide file encrypt -g myGroup -u john@example.com,mike@example.com path/to/file

    Encrypt the provided file and write the decrypted bytes to 'other/path/to/encrypted.iron'.
        $ ironhide file encrypt path/to/file -o other/path/to/encrypted.iron

    Encrypt the provided file and write the decrypted bytes to 'path/to/file.txt.iron' while also deleting
    the original unencrypted 'path/to/file.txt' file.
        $ ironhide file encrypt path/to/file.txt -d

    Encrypt all of the JSON files in the current directory and write them out to '.iron' files.
        $ ironhide file encrypt *.json

    Encrypt the provided file and write the encrypted content to stdout.
        $ ironhide file encrypt path/to/file -o -

    Encrypt stdin and write the encrypted results to stdout in base64.
        $ echo \"my secret\" | ironhide file encrypt -s -o - | base64 -e
";

#[derive(Parser)]
#[clap(after_help = EXAMPLES)]
/// Encrypt a file or list of files to yourself and optionally to other users or groups. By default, the input file is
/// unchanged and the output uses the same filename with a '.iron' extension added.
pub struct Encrypt {
    /// Delete the unencrypted source file(s) after successful encryption.
    #[clap(short, long, num_args = 0)]
    delete: bool,
    /// Path of file(s) to encrypt.
    #[clap(value_parser = clap::value_parser!(PathBuf), num_args = 1.., required = true)]
    files: Vec<PathBuf>,
    /// Encrypt the file(s) to the given groups. Multiple groups should be comma separated.
    /// Can refer to a group by ID or by name. Indicate IDs by prefixing with 'id^' e.g. 'id^groupID'.
    #[clap(value_parser = util::group_identifier_from_string, short, long, use_value_delimiter = true, value_delimiter = ',')]
    groups: Vec<Either<GroupName, GroupId>>,
    /// Path to location of file which contains keys to use for this operation. Overrides using default key file from
    /// '~/.iron' directory.
    #[clap(value_parser = clap::value_parser!(PathBuf), short, long)]
    keyfile: Option<PathBuf>,
    /// Filename where encrypted file will be written. Only allowed if a single file is
    /// being encrypted.
    /// Use '-o -' to write encrypted file content to stdout, but fair warning, the output is binary and not ASCII.
    #[clap(value_parser = clap::value_parser!(PathBuf), short, long)]
    out: Option<PathBuf>,
    /// Read data to encrypt from stdin. If used, no source files should be provided as
    /// arguments and you must use the '-o' flag.
    #[clap(short, long, num_args = 0, conflicts_with = "files", requires = "out")]
    stdin: bool,
    /// Encrypt the file(s) to a comma-separated list of user emails. Files are
    /// automatically encrypted to the logged-in user.
    #[clap(value_parser = parse_user_id, short, long, use_value_delimiter = true, value_delimiter = ',', required = false)]
    users: Vec<UserId>,
}

fn parse_user_id(s: &str) -> Result<UserId, IronOxideErr> {
    UserId::try_from(s)
}

impl util::GetKeyfile for Encrypt {
    fn get_keyfile(&self) -> Option<&PathBuf> {
        self.keyfile.as_ref()
    }
}

fn display_group_name_and_id(meta: &GroupMetaResult) -> String {
    format!(
        "{} ({})",
        meta.name()
            .map(|n| n.name().as_str())
            .unwrap_or_else(|| "NO_NAME"),
        meta.id().id()
    )
}

pub fn encrypt_files(
    sdk: &BlockingIronOxide,
    Encrypt {
        delete,
        files,
        groups,
        out,
        stdin,
        users,
        ..
    }: Encrypt,
) -> Result<(), String> {
    if stdin {
        let mut file: Vec<u8> = vec![];
        match io::stdin().read_to_end(&mut file) {
            Ok(_) => {
                match encrypt_file(&out, None, sdk, &groups, &users, file, delete) {
                    Ok(_) => (),
                    Err(e) => {
                        util::println_paint(Paint::red(format!("Error encrypting file: {}", e)))
                    }
                };
            }
            Err(e) => util::println_paint(Paint::red(format!("Error reading stdin: {}", e))),
        }
    } else if out.is_some() && files.len() > 1 {
        util::println_paint(Paint::red(
            "Cannot use '-o' flag with multiple files.".to_string(),
        ));
    } else {
        let result = act_on_all_files(
            &files,
            |infile| -> Result<EncryptResultWithResolved, String> {
                let file = fs::read(infile).map_err(|e| {
                    format!(
                        "Provided path '{}' doesn't exist or is not readable: {e}",
                        infile.display()
                    )
                })?;
                let encrypt_result =
                    encrypt_file(&out, Some(infile), sdk, &groups, &users, file, delete)?;
                if files.len() == 1 {
                    util::println_paint(Paint::green(format!(
                        "Encrypted file successfully written to {}.",
                        encrypt_result.output_log.clone()
                    )));
                }
                Ok(encrypt_result)
            },
            "encrypted",
        )
        .map_err(|(e, maybe_success)| {
            maybe_success.map(|success| print_resolved_grants(success));
            e
        })?;
        result.map(|encrypt_result| print_resolved_grants(encrypt_result));
    }

    Ok(())
}

fn print_resolved_grants(encrypt_result: EncryptResultWithResolved) {
    util::println_paint(Paint::green(
        "\nSuccessfully encrypted to the following users and groups:".to_string(),
    ));
    let zipped = encrypt_result
        .resolved_users
        .into_iter()
        .zip_longest(encrypt_result.resolved_groups);
    let mut table = table!([Fbb=>"Users", "Groups"]);
    zipped.for_each(|either_or_both| {
        let row = match either_or_both {
            EitherOrBoth::Both(ref user, ref group) => {
                (user.id(), display_group_name_and_id(group))
            }
            EitherOrBoth::Left(ref user) => (user.id(), "".to_string()),
            EitherOrBoth::Right(ref group) => ("", display_group_name_and_id(group)),
        };
        table.add_row(Row::new(vec![cell!(Fw -> row.0), cell!(Fw -> row.1)]));
    });
    table.printstd();
}

fn encrypt_file(
    out: &Option<PathBuf>,
    infile: Option<&PathBuf>,
    sdk: &BlockingIronOxide,
    groups: &[Either<GroupName, GroupId>],
    users: &[UserId],
    file: Vec<u8>,
    delete: bool,
) -> Result<EncryptResultWithResolved, String> {
    let (groups_by_name, groups_by_id) = get_group_maps(sdk);
    let group_ids = convert_group_names_to_ids(groups, &groups_by_name);
    let users_or_groups = util::collect_users_and_groups(users, &group_ids);
    let encrypt_result = encrypt_bytes_to_file(sdk, &file, &users_or_groups, out, infile)?;
    let (resolved_users, maybe_resolved_groups): (Vec<UserId>, Vec<Option<GroupMetaResult>>) =
        encrypt_result
            .grants
            .into_iter()
            .partition_map(|u_or_g| match u_or_g {
                UserOrGroup::User { id } => Either::Left(id),
                UserOrGroup::Group { id } => Either::Right(groups_by_id.get(&id).cloned()),
            });
    let resolved_groups = maybe_resolved_groups.into_iter().flatten().collect_vec();
    if delete {
        match infile {
            Some(infile) => {
                if fs::remove_file(infile).is_err() {
                    util::println_paint(Paint::yellow(format!(
                        "Unable to delete source file '{}' as it is not writable.",
                        &infile.display()
                    )))
                }
            }
            None => util::println_paint(Paint::yellow(
                "Unable to delete source as it was a stream.".to_string(),
            )),
        }
    }
    Ok(EncryptResultWithResolved {
        resolved_users,
        resolved_groups,
        output_log: encrypt_result.output_log,
    })
}

/// Validate that the output path provided by the user can be used for encryption. If no path is provided,
/// will append ".iron" to the input filename. Returns an Err if the input file ends with "..".
fn validate_encrypt_output_path(
    maybe_output: Option<PathBuf>,
    infile: Option<&PathBuf>,
) -> Result<(Box<dyn Write>, String), String> {
    let output: (Box<dyn Write>, String) = match maybe_output {
        // User specified an output path.
        Some(desired) => {
            // User specified a directory for output
            if desired.is_dir() && infile.is_some() {
                let mut filename = infile
                    // unwrap here is safe because of the .is_some check above
                    .unwrap()
                    .file_name()
                    .ok_or_else(|| "Invalid input file".to_string())?
                    .to_os_string();
                filename.push(".iron");
                let mut desired_dir = desired;
                desired_dir.push(filename);
                // if desired_dir.exists() {
                //     return Err("BLOW UP".to_string())
                // }
                (
                    Box::new(
                        OpenOptions::new()
                            .create_new(true)
                            .append(true)
                            .open(desired_dir.clone())
                            .map_err(|e| {
                                format!(
                                    "Couldn't create a file in the desired output directory '{}': {e}",
                                    desired_dir.display()
                                )
                            })?
                    ),
                    desired_dir.to_string_lossy().to_string(),
                )
            } else if desired == PathBuf::from("-") {
                (Box::new(io::stdout()), "stdout".to_string())
            } else {
                (
                    Box::new(
                        OpenOptions::new()
                            .create_new(true)
                            .append(true)
                            .open(desired.clone())
                            .map_err(|e| {
                                format!(
                                    "Couldn't create a file at the desired output path '{}': {e}",
                                    desired.display()
                                )
                            })?,
                    ),
                    desired.to_string_lossy().to_string(),
                )
            }
        }
        // User didn't specify an output path. Add ".iron" to the input path.
        None => match infile {
            Some(infile) => {
                let input_iron = infile.display().to_string() + ".iron";
                (
                    Box::new(
                        OpenOptions::new()
                            .create_new(true)
                            .append(true)
                            .open(PathBuf::from(input_iron.clone()))
                            .map_err(|e| {
                                format!("Couldn't create a file at the default output path '{input_iron}': {e}")
                            })?,
                    ),
                    input_iron,
                )
            }
            None => unreachable!(), // If stdin is used, output path is required, this is unreachable.
        },
    };
    Ok(output)
}

struct EncryptResult {
    grants: Vec<UserOrGroup>,
    output_log: String,
}

#[derive(Clone)]
struct EncryptResultWithResolved {
    resolved_users: Vec<UserId>,
    resolved_groups: Vec<GroupMetaResult>,
    output_log: String,
}

/// Encrypt the provided file to the `users_or_groups`. The file will also be granted to the calling user.
/// The bytes of the encrypted file will be written to `output_path`.
fn encrypt_bytes_to_file(
    sdk: &BlockingIronOxide,
    file: &[u8],
    users_or_groups: &[UserOrGroup],
    outfile: &Option<PathBuf>,
    infile: Option<&PathBuf>,
) -> Result<EncryptResult, String> {
    let grants = ExplicitGrant::new(true, users_or_groups);
    let opts = DocumentEncryptOpts::new(None, None, EitherOrBoth::Left(grants));
    let encrypt_result = sdk.document_encrypt(file.to_vec(), &opts)?;
    let (mut output_writer, output_log) = validate_encrypt_output_path(outfile.clone(), infile)?;

    output_writer
        .write_all(encrypt_result.encrypted_data())
        .map_err(|e| format!("Couldn't write encrypted file: {e}"))?;

    Ok(EncryptResult {
        grants: encrypt_result.grants().to_vec(),
        output_log,
    })
}
